src/foam/nanos/auth/FilterBySpidService.js

```diff
@@ -0,0 +1,74 @@
+/**
+ * @license
+ * Copyright 2021 The FOAM Authors. All Rights Reserved.
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+foam.CLASS({
+   package: 'foam.nanos.auth',
+   name: 'FilterBySpidService',
+   extends: 'foam.nanos.auth.ProxyUserLocatorService',
+   flags: ['java'],
+
+  documentation: `Restrict login to the url that matches the spid of the user.`,
+
+
+  javaImports: [
+    'foam.dao.DAO',
+    'foam.nanos.auth.AuthenticationException',
+    'foam.nanos.auth.User',
+    'foam.nanos.theme.Theme',
+
+    'static foam.mlang.MLang.*'
+  ],
+
+  properties: [
+    {
+      class: 'String',
+      name: 'superSpid',
+      documentation: 'Set spid of a theme that is accessible to all users',
+      value: "",
+    }
+  ],
+
+  methods: [
+    {
+      name: 'getUser',
+      documentation: 'Helper logic function to reduce code duplication',
+      type: 'User',
+      args: [
+        {
+          name: 'x',
+          type: 'Context'
+        },
+        {
+          name: 'identifier',
+          type: 'String'
+        },
+        {
+          name: 'password',
+          class: 'String'
+        }
+      ],
+      javaCode: `
+        DAO userDAO = (DAO) getX().get("localUserDAO");
+
+        userDAO = userDAO
+          .where(
+          AND(
```

Comment:
AND-ing OR's can't be indexed, so will do a full-table scan. Which is okay now, but won't be as good when we have millions of users. Better if you can reformulate your query so that the top-level operator is an OR, or make separate find's.

Reply:
it's good to know. Thank you!